core/src/main/java/hudson/model/AbstractProject.java

```diff
@@ -983,16 +985,139 @@ protected R loadBuild(File dir) throws IOException {
      * @see TransientProjectActionFactory
      */
     @SuppressWarnings("deprecation")
+    @Nonnull
     @Override
     public List<Action> getActions() {
         // add all the transient actions, too
-        List<Action> actions = new Vector<Action>(super.getActions());
+        List<Action> actions = new Vector<>(super.getActions());
         actions.addAll(transientActions);
         // return the read only list to cause a failure on plugins who try to add an action here
         return Collections.unmodifiableList(actions);
     }
 
-    // TODO implement addAction, addOrReplaceAction, removeAction, removeActions, replaceActions
+    // by historical reasons Actionable.getActions() is dealing with persistent modifiable list
+    // while child classes lock them to be unmodifiable, so addAction, addOrReplaceAction, removeActions...
+    // duplicating code. no way..
+    @SuppressWarnings("deprecation")
+    public List<Action> getPersistentActions() {
+        return super.getActions();
+    }
+
+    /**
+     * @see Actionable#addAction(hudson.model.Action)
+     */
+    @SuppressWarnings({"ConstantConditions","deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    @Override
+    public void addAction(@Nonnull Action a) {
+        if (a == null) {
+            throw new IllegalArgumentException("Action must be non-null");
+        }
+        super.getActions().add(a);
+    }
+
+    /**
+     * @see Actionable#addOrReplaceAction(hudson.model.Action)
+     */
+    @SuppressWarnings({"ConstantConditions", "deprecation"})
+    @SuppressFBWarnings("RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE")
+    public boolean addOrReplaceAction(@Nonnull Action a) {
```

Comment:
Return the replaced action instead?

Reply:
That's all copy-paste from Actionable without changing behaviour